### ISSUE TITLE
[Bug] Missing items in Trakt Movies Collection

### ISSUE THREAD

**Issue by burekas7** [elgatito/plugin.video.elementum#680]
Hi,

I recently added a new movie to the Trakt Movies Collection.
But when enter to the movies collection in Elementum, it isn't there.
(In Trakt site it's there, and also in other addons that show the Trakt Collection)

Instead of getting 166 movies from Movies Collection, it still show 165.
The last movie that was added is "The Croods: A New Age"
It doesn't show new items that were added.

There are also errors notification regarding the TMDB, maybe there is a connection to this problem.
Maybe something with the api key.
`[plugin.video.elementum] WARN  tmdb    CheckAPIKey      TMDB API key failed:`

Kodi debug log:
https://paste.ubuntu.com/p/pTr6Npz5p9

Elementum: 0.1.66
Kodi 17.6
Win 10, 64bit

**Comment by burekas7** [comment]
@elgatito 

Some finding regarding this problem.

1) I found that pressing on "Clear Trakt cache" in the Settings refreshes the collection and the last item that was added is shown.
By the wat, even after restarting kodi the Trakt Collection isn't refreshed in Elementum.
Only after few days / uses the lists were updated or by some action that I did in Elementum that make it to be refreshed and updating the cache.

2) The problem is depends on from where you added the movie to the collection:
- When adding it through the Elementum contextmenu inside Elementum addon - It's ok and lists are refreshed.
- When adding it through external addon / or from Trakt site - It's not ok and lists are not refreshed.

So my conclusion is:
Refreshing the Trakt lists on each time you enter it.
I understand that it's much faster to open Trakt lists from cache as it now, but the lists doesn't show the updated data when you manipulate it from an external source.

**Comment by antonsoroko** [comment]
@burekas7 
as i can see in the code sync with trakt is done:
1) on plugin start, after `library_update_delay` (in order not to stress system much right after kodi start) **seconds** (default=300). so 10 seconds delay seems to be ok i think.

2) in a background, every trakt_sync_frequency_min **minutes** (default=5). It gets last activities (see https://trakt.docs.apiary.io/#reference/sync/last-activities/get-last-activity) from trakt and update things correspondingly. 
maybe you set this value to some big value. or there were some trakt server issues.

it works fine for me, but sure, it can be improved by also checking if list was updated right before we render list (currently, as i mentioned, check is done only in a backgrounds with interval).

also worth to mention that "Clear Trakt cache" is quite bad since it will force elementum to sync again **all** you stuff from trakt from scratch.

**Comment by antonsoroko** [comment]
@burekas7 

in your log

>2022-02-03 02:38:20.975 T:20516 WARNING : [plugin.video.elementum] TraktSyncFrequencyMin: 50,

so you have to wait 50 minutes for sync to happen. default is 5. set it it 5 and test.

**Comment by burekas7** [comment]
@antonsoroko 
Doesn't work.

I remove one of the movies from my Trakt Collection.
And the movie is still there, even after "Synchronize frequency of Trakt" should be done (I have changed it to 5 minutes, and still the same)
* By the way, does the action button of "Update Trakt" manually should be done the same as this option?

**Comment by antonsoroko** [comment]
>By the way, does the action button of "Update Trakt" manually should be done the same as this option?

yes. it runs this function https://github.com/elgatito/elementum/blob/6aea37dab272b523d6ec42dd75ef123031e388d6/library/trakt.go#L27 which is also runs every "Synchronize frequency of Trakt" minutes.

if there was any activity - then it actually does update of appropriate thing (like lists or collections).

try to use this API https://trakt.docs.apiary.io/#reference/sync/last-activities/get-last-activity  to see if trakt avtually recorded the fact that you added something into collection. for me it works. for you - i do not know.

>Doesn't work.

it works for me. i added movie, waited 5 minutes, it was added, i removed it, waited 5 minutes, it was removed.
i am talking about "elementum->movies->trakt collections".

maybe you do it differently than me or you have different settings (this is my bet).

compare your settings and mine: [elementum settings.zip](https://github.com/elgatito/plugin.video.elementum/files/13762408/elementum.settings.zip)

**Comment by burekas7** [comment]
@antonsoroko 
There are a lot of difference between us :)
What should I need to focus on?

The "Update Trakt" doesn't update the Collection list. (Trakt website -> 182 Movies, Elementum -> 183 Movies)
Only if I will do a clear cache, but I don't want to of course.

Does the log can be helpful here?

**Comment by antonsoroko** [comment]
@burekas7 can please you try to install new kodi separately?
on windows you can use "portable" version kodi i believe.

and then please test this issue on new kodi+elementum **with default settings.**

and if you will still have issue - then i will try to add some debug logging to "collection sync" function and i will make a test build - thus log file will be useful. 
but i need you to test it on fresh installation first.

**Comment by burekas7** [comment]
@elgatito @antonsoroko 

Ok, this is one of the strangest thing.
I found how to reproduce the bug:

1. Kodi: Install a fresh Kodi with a fresh Elementum, authorize your Trakt user.
2. Trakt website: Add a movie to your collection
3. Kodi: Open Elementum Settings -> Trakt -> Update Trakt
4. Kodi: Go back to Elementum - Movies - Trakt> My Collection
5. Check if the movie was added.
6. Repeate steps 2 -> 5, but this time remove a movie and check if it was removed. then go to 7.
7. Kodi: Open Elementum Settings -> Trakt -> Disable "Sync items from collections into Kodi library"
8. Repeate steps 2 -> 5, but this time you will see the bug, the movie won't be added to the list.
9. Even though you will enable back the "Sync items from collections into Kodi library" settings, and make a new change in Trakt website, it won't fix it. The bug remains forever until you reset the settings to default.

* By the way, I checked all the other settings in the sections there of "Synchronization", and this is the only setting that makes the bug once you change it at the first time.

The problem is that I want this setting to be disabled, so I need to touch it.

Why does it happen?

**Comment by antonsoroko** [comment]
@burekas7 

>Repeate steps 2 -> 5, but this time you will see the bug, the movie won't be added to the list.

this is not a bug :-) you disabled sync of collection and thus collection is not synced. this is cause and effect, see https://en.wikipedia.org/wiki/Causality.
unless you wanted to write that sentence somehow differently.

>Even though you will enable back the "Sync items from collections into Kodi library" settings, and make a new change in Trakt website, it won't fix it. 

it **works** for me. after i enabled back "sync of collection" i had to make another change (because my previous "activity" provided by trakt api was ignored by elementum since "sync of collection" option was disabled, which is logical) - everything synced.

open your log file and check for likes like that:
```
warning <general>: [plugin.video.elementum] INFO  library      ▶ RefreshTrakt     Running Trakt sync
warning <general>: [plugin.video.elementum] WARN  library      ▶ RemoveShow       Directory /home/user/.kodi/elementum_library/Shows/Childhood's End (2015) removed from disk
warning <general>: [plugin.video.elementum] WARN  library      ▶ RemoveShow       Конец детства removed from library
warning <general>: [plugin.video.elementum] INFO  library      ▶ SyncShowsList    Shows list (collection) removed 1 items

warning <general>: [plugin.video.elementum] INFO  library      ▶ RefreshTrakt     Running Trakt sync
warning <general>: [plugin.video.elementum] WARN  library      ▶ RemoveMovie      Directory /home/user/.kodi/elementum_library/Movies/Harry Potter and the Philosopher's Stone (2001) removed from disk
warning <general>: [plugin.video.elementum] WARN  library      ▶ RemoveMovie      Гарри Поттер и философский камень removed from library
warning <general>: [plugin.video.elementum] INFO  library      ▶ SyncMoviesList   Movies list (collection) removed 1 items
```



>The problem is that I want this setting to be disabled, so I need to touch it, but if I'm touching it I make the bug.

i do not understand you. if you do not want collection to be synced - then do **not** enabled it.
if you want collection to be synced - then enable.
it is not possible to sync collection without enabling options called "sync collection", see https://en.wikipedia.org/wiki/Mutual_exclusivity

**Comment by burekas7** [comment]
@antonsoroko 
Read the setting again "into Library"
I don't want it to be sync with my Library, but I do want it to be synced inside Elementum (Trakt -> My Collection)
These are two different things.

Regarding the second part, when I enable it it doesn't sync again once I disabled it on the first time.

**Comment by antonsoroko** [comment]
@burekas7 i found it.
and it is actually not only for collections, but also for watchlist and userlists too - those lists (technically all those are lists, so i will can them lists) will not be updated.
see this check:
https://github.com/elgatito/elementum/blob/6aea37dab272b523d6ec42dd75ef123031e388d6/library/trakt.go#L651
https://github.com/elgatito/elementum/blob/6aea37dab272b523d6ec42dd75ef123031e388d6/library/trakt.go#L672
https://github.com/elgatito/elementum/blob/6aea37dab272b523d6ec42dd75ef123031e388d6/library/trakt.go#L801

but then in `SyncMoviesList` (and in other similar functions) we do make request to trakt (thus we update data for "view" functions which are used when you open those lists) and also write to library.
so we need to split those 2 actions in all functions related to collections/watchlist/userlists.

**Comment by burekas7** [comment]
@antonsoroko @elgatito 
1. Ok, so we need to fix it so there is no dependence with Kodi Library  regarding the sync with any Trakt list.
(This means anyone of these setting of "Sync items from ... into Kodi library")
Is it complicated?
2. I guess it should be the same also with `SyncShowsList`
3. For me it does. Enable it again, confirm settings, then "Update Trakt" again in the settings. It doesn't refresh the Trakt list of My Collection. The specific movie that removed is still there unless I reset to defaults as I described etc..

**Comment by antonsoroko** [comment]
@burekas7 
3. are you sure that after you enabled back the option **you also made appropriate change (like if you want to force update of collection - then make change to collection) on trakt website to force update?** just clicking "update trakt" will do nothing, as i explained it above:
>i had to make another change (because my previous "activity" provided by trakt api was ignored by elementum since "sync of collection" option was disabled, which is logical) and after that - everything synced.

**Comment by antonsoroko** [comment]
@burekas7 
1. it depends on how beautiful code we want to have :-) i made more-less simple fix - try it https://disk.yandex.ru/d/eOXSo5Mgeb9WiA
note: this build does not have the latest changes about counting, since corresponding PR was not merged yet by elagatito. But to test in clean installation - should be ok. And you always can download that build with counting changes.

**Comment by burekas7** [comment]
> 1. it depends on how beautiful code we want to have :-) i made more-less simple fix - try it https://disk.yandex.ru/d/eOXSo5Mgeb9WiA
>    note: this build does not have the latest changes about counting, since corresponding PR was not merged yet by elagatito. But to test in clean installation - should be ok. And you always can download that build with counting changes.

Thanks. It's working.

Trakt > My Collection is updated even when "Sync items from collections into Kodi library" is disabled.

> 3. are you sure that after you enabled back the option you also made appropriate change (like if you want to force update of collection - then make change to collection) on trakt website to force update? just clicking "update trakt" will do nothing

Yes, I didn't make a change after that.
I thought that it checks difference with the list of Trakt and update, I forgot you mentioned that it relies on Trakt activity.
But according to your fix, it's probably not relevant anymore :)

**Comment by antonsoroko** [comment]
@burekas7 nice. i hope we done here. so now we need to wait for PR to be merged.

**Comment by elgatito** [comment]
@burekas7 You can try the fix with v0.1.99.

Ideally it should be a decider in caching layer to determine cache expiration date based on "freshness" of a Movie/Show/Season/Episode. If something is pretty fresh or airing - we should not cache for more than 1 day.

**Comment by burekas7** [comment]
> @burekas7 You can try the fix with v0.1.99.
> 
> Ideally it should be a decider in caching layer to determine cache expiration date based on "freshness" of a Movie/Show/Season/Episode. If something is pretty fresh or airing - we should not cache for more than 1 day.

Doesn't work with v0.1.99
Are you sure that you use @antonsoroko solution?
With his build it worked.

**Comment by antonsoroko** [comment]
@burekas7 strange, it works for me. provide a log file please.

**Comment by burekas7** [comment]
@elgatito @antonsoroko 

Yes, you right.
I forgot to make a change (activity) so that it would also detect in Elementum and will cause the list to be refreshed

Approved with v0.1.99.
Close here.
Thanks
